src/assignment-5/CollateralizedVault.sol

```diff
@@ -0,0 +1,168 @@
+
+    /// @dev Number of decimals for collateral token
+    uint8 public immutable collateralDecimals;
+
+    /// @dev Number of decimals for oracle price
+    uint8 public immutable oracleDecimals;
+
+    /******************
+     * State variables
+     ******************/
+
+    /// @notice deposited collateral per user in the vault, of `collateral` token
+    mapping(address => uint256) public deposits;
+
+    /// @notice Mapping of current debt per user, of `underlying` token
+    mapping(address => uint256) public borrows;
+
+
+    error TooMuchDebt();
+    error NotEnoughCollateral();
+    error UserDebtIsSufficientlyCollateralized();
+
+    /// @notice Initalizes a new CollateralizedVault
+    /// @param underlying_ ERC20 token which can be borrowed from the vault
+    /// @param collateral_ ERC20 token which can be used as collateral
+    /// @param oracle_ Chainlink price feed of underlying / collateral
+    constructor(address underlying_, address collateral_, address oracle_) {
+        underlying = IERC20WithDecimals(underlying_);
+        collateral = IERC20WithDecimals(collateral_);
+        oracle = AggregatorV3Interface(oracle_);
+
+        // Gas optimization: save decimals in contract instead of reading from external contract
+        underlyingDecimals = underlying.decimals();
+        collateralDecimals = collateral.decimals();
+        oracleDecimals = oracle.decimals();
+    }
+
+    /// @notice Deposits additional collateral into the vault
+    /// @param collateralAmount amount of `collateral` token to deposit
+    function deposit(uint256 collateralAmount) public {
+        deposits[msg.sender] += collateralAmount;
+
+        collateral.transferFrom(msg.sender, address(this), collateralAmount);
+    }
+
+    /// @notice Borrows `underlying` token from the Vault
+    ///     Only allowed to borrow up to the value of the collateral
+    /// @param amount The amount of `underlying` token to borrow
+    function borrow(uint256 amount) public {
+        if (getRequiredCollateral(borrows[msg.sender] + amount) > deposits[msg.sender]) {
+            revert NotEnoughCollateral();
+        }
+
+        borrows[msg.sender] += amount;
+
+        underlying.transfer(msg.sender, amount);
+    }
+
+    /// @notice Pays back an open debt
+    /// @param amount The amount of `underlying` token to pay back
+    function repay(uint256 amount) public {
+        borrows[msg.sender] -= amount;
+
+        underlying.transferFrom(msg.sender, address(this), amount);
+    }
+
+    /// @notice Withdraws part of the collateral
+    ///     Only allowed to withdraw as long as the collateral left is higher in value than the debt
+    /// @param collateralAmount The amount of `collateral` token to withdraw
+    function withdraw(uint256 collateralAmount) public {
+        uint256 requiredCollateral = getRequiredCollateral(borrows[msg.sender]);
+
+        if (deposits[msg.sender] - collateralAmount < requiredCollateral) {
```

Comment:
this is a pattern i learned from maple-labs, i think its cool and @alberto hasnt scolded me for it yet lol
```
   if ((deposits[msg.sender] -= collateralAmount) < requiredCollateral) { ...
```
then u can delete line 115

i think its cleaner

Reply:
@devtooligan 
mmmmm to be honest, I don't love this style.

mixing statements & expressions is quite confusing to read IMO
it can start to get really confusing in `c` style expressions like the difference between `i++` and `++i`.

I would do something like this only if it's saving gas or has some other benefit.

also: for the checks-effects-interactions pattern, this would make an "effect" ran before the "check"

Reply:
cool i dont feel strongly about it

Reply:
I wouldn't scold anyone, I'm also guilty sometimes of compacting code too much (love ternary expressions!). I could get used to the pattern @devtooligan suggested, but I agree that it is confusing.

In short. You can choose. I think simplicity is probably best, I guess.